### PR TITLE
Initial pass with top level form selection and deletion

### PR DESCRIPTION
**Pull request by Olical**
Could maybe do with more methods and tests, but this is already working and the tests are passing. I haven't checked visual selections yet, I presume I will need to add more functions and mappings for that.

The general tree sitter / selection code seems to be working though. I might also refactor this a bit to reuse code between the "current form" and "top level form" functions. There's a lot of overlap there but I thought I'd take the simple copy/paste approach first.

Also I went with "top level" to describe this concept, in my tooling I call it "root" but you already have terms for root and I didn't want to muddy those waters.

### CODE REVIEW

**Comment by Olical** [comment]
Okay, code deduplicated and another test for visual selections added. Let me know what you think / what you'd like to change! And feel free to just take this and run with it if you want to make changes of course 😄 

I hope this helps!

See #12 for more information about this feature. (for future readers / people without context)

**Review by julienvincent — APPROVED** [review]
Thank you for taking the time to add this! This is a great improvement with a very clean implementation.

I like the "top level form" terminology too :)

File: lua/nvim-paredit/utils/traversal.lua [hunk 1]
```diff
@@ -139,4 +139,34 @@ function M.find_root_element_relative_to(root, child)
...
+  --   -- Node X
+  --   --- Sub-node 1
+  --   - Branch B
+  --   -- Node Y
+  --   --- Sub-node 2
+  --   --- Sub-node 3
+
+  -- If we call this function on "Sub-node 2" we expect "Branch B" to be
+  -- returned, the top level one below the document itself. We know which
+  -- node is the document because it lacks a parent, just like Batman.
```
Comment (julienvincent) [inline lua/nvim-paredit/utils/traversal.lua:154]: 🦇

**Comment by Olical** [comment]
I'm glad you like it! And thanks for building this great plugin, it's a really nice framework to extend.

(so thankful that this is lua, I can't bring myself to contribute / work with VimL these days at all 😬 )

**Comment by julienvincent** [comment]
> I'm glad you like it! And thanks for building this great plugin, it's a really nice framework to extend.

You're welcome 🤗 It's the least we can do in exchange for Conjure ;)

> (so thankful that this is lua, I can't bring myself to contribute / work with VimL these days at all 😬 )

I know! My aversion to VimL was a large factor in why this project even happened! Haha

**Pull request merged**